wallet/store.go

```diff
@@ -1,164 +1,245 @@
+		if k.salt == nil {
+			log.Fatalf("Salt must be set.")
+		}
+		if k.key != nil {
+			log.Fatalf("Can only generate key once.")
+		}
+		iterations := k.iterations
+		if iterations == 0 {
+			iterations = Pbkdf2Iterations
+		}
+		k.key = pbkdf2.Key(
+			password,
+			k.salt,
+			iterations,
+			EncKeyLen,
+			Pbkdf2HashFunc,
+		)
+		k.pw = password
+	}
 }
 
-type ExportableWallet struct {
-	// EncryptedWallet is the encrypted wallet data in base58 encoding.
-	EncryptedWallet string `json:"encrypted_wallet"`
-	// Salt is the salt used to derived the wallet's encryption key in base58 encoding.
-	Salt string `json:"salt"`
-	// Nonce is the nonce used to encrypt the wallet in base58 encoding.
-	Nonce []byte `json:"nonce"`
-}
+// https://cheatsheetseries.owasp.org/cheatsheets/Secrets_Management_Cheat_Sheet.html#71-encryption-types-to-use
+func (k *WalletKey) encrypt(plaintext []byte) (ciphertext []byte, nonce []byte, err error) {
+	block, err := aes.NewCipher(k.key)
+	if err != nil {
+		return
+	}
 
-type WalletStore struct {
-	wk WalletKey
+	// Using default options for AES-GCM as recommended by the godoc.
+	// For reference, NonceSize is 12 bytes, and TagSize is 16 bytes:
+	// https://cs.opensource.google/go/go/+/refs/tags/go1.19.2:src/crypto/cipher/gcm.go;l=153-158
+	aesgcm, err := cipher.NewGCM(block)
+	if err != nil {
+		return
+	}
+	hash := hmac.New(sha512.New, k.key)
+	nonce = hash.Sum(plaintext)[:aesgcm.NonceSize()]
+
+	ciphertext = aesgcm.Seal(nil, nonce, plaintext, nil)
+	return
 }
 
-func NewStore(wk *WalletKey) *WalletStore {
-	return &WalletStore{
-		wk: *wk,
+func (k *WalletKey) decrypt(ciphertext []byte, nonce []byte) (plaintext []byte, err error) {
+	block, err := aes.NewCipher(k.key)
+	if err != nil {
+		return
+	}
+	aesgcm, err := cipher.NewGCM(block)
+	if err != nil {
+		return
 	}
-}
 
-// Just a quick storage of the encrypted mnemonic for now.
-// TODO: add metadata, decide what else should actually go in the "core wallet"
+	plaintext, err = aesgcm.Open(nil, nonce, ciphertext, nil)
+	return
+}
 
-func (s *WalletStore) Open(file *os.File) (w *Wallet, err error) {
+func (k *WalletKey) Open(file *os.File) (w *Wallet, err error) {
 	jsonWallet, err := os.ReadFile(file.Name())
 	if err != nil {
 		return nil, err
 	}
-	ew := &ExportableWallet{}
+	ew := &EncryptedWalletFile{}
 	if err = json.Unmarshal(jsonWallet, ew); err != nil {
 		return nil, err
 	}
-	s.wk.salt = base58.Decode(ew.Salt) // Replace auto-generated salt with the one from the wallet file.
-	encWallet := base58.Decode(ew.EncryptedWallet)
-	decMnemonic := s.wk.decrypt(encWallet, ew.Nonce)
-	w = WalletFromMnemonic(string(decMnemonic))
-	return w, nil
+	// this returns an EOF error for some reason
+	//if err = json.NewDecoder(file).Decode(ew); err != nil {
+	//	return
+	//}
+
+	// set the salt, and warn if it's different
+	var salt [Pbkdf2SaltBytesLen]byte
+	if _, err = hex.Decode(salt[:], []byte(ew.Secrets.KDFParams.Salt)); err != nil {
+		return
+	}
```

Comment:
If `KDFParams.Salt` is also a `hexEncodedCiphertext` you can just do this instead:

```suggestion
	copy(salt[:], ew.Secrets.KDFParams.Salt)
```